langchain4j/src/main/java/dev/langchain4j/data/document/splitter/DocumentByMarkdownHeaderSplitter.java

```diff
@@ -0,0 +1,159 @@
+     */
+    public DocumentByMarkdownHeaderSplitter(
+            Set<MarkdownHeaderLevel> headerLevels,
+            String joinDelimiter,
+            int maxSegmentSizeInChars,
+            int maxOverlapSizeInChars,
+            DocumentSplitter subSplitter) {
+        super(maxSegmentSizeInChars, maxOverlapSizeInChars, null, subSplitter);
+        this.joinDelimiter = joinDelimiter;
+        this.pattern = buildPattern(headerLevels);
+    }
+
+    /**
+     * Constructor for DocumentByMarkdownHeaderSplitter.
+     *
+     * @param headerLevels           The levels of headers to split by.
+     * @param joinDelimiter          The delimiter to use when joining segments.
+     * @param maxSegmentSizeInTokens The maximum size of a segment in tokens.
+     * @param maxOverlapSizeInTokens The maximum size of overlap between segments in tokens.
+     * @param tokenizer              The tokenizer to use for splitting the text into tokens.
+     */
+    public DocumentByMarkdownHeaderSplitter(
+            Set<MarkdownHeaderLevel> headerLevels,
+            String joinDelimiter,
+            int maxSegmentSizeInTokens,
+            int maxOverlapSizeInTokens,
+            Tokenizer tokenizer) {
+        super(maxSegmentSizeInTokens, maxOverlapSizeInTokens, tokenizer, null);
+        this.joinDelimiter = joinDelimiter;
+        this.pattern = buildPattern(headerLevels);
+    }
+
+    /**
+     * Constructor for DocumentByMarkdownHeaderSplitter.
+     *
+     * @param headerLevels           The levels of headers to split by.
+     * @param joinDelimiter          The delimiter to use when joining segments.
+     * @param maxSegmentSizeInTokens The maximum size of a segment in tokens.
+     * @param maxOverlapSizeInTokens The maximum size of overlap between segments in tokens.
+     * @param tokenizer              The tokenizer to use for splitting the text into tokens.
+     * @param subSplitter            The sub-splitter to use for further splitting the segments.
+     */
+    public DocumentByMarkdownHeaderSplitter(
+            Set<MarkdownHeaderLevel> headerLevels,
+            String joinDelimiter,
+            int maxSegmentSizeInTokens,
+            int maxOverlapSizeInTokens,
+            Tokenizer tokenizer,
+            DocumentSplitter subSplitter) {
+        super(maxSegmentSizeInTokens, maxOverlapSizeInTokens, tokenizer, subSplitter);
+        this.joinDelimiter = joinDelimiter;
+        this.pattern = buildPattern(headerLevels);
+    }
+
+    /**
+     * Splits the given text by the pattern.
+     *
+     * @param text The text to split.
+     * @return An array of strings, each representing a segment of the text.
+     */
+    @Override
+    public String[] split(String text) {
+        return splitTextByPattern(text);
+    }
+
+    @Override
+    public String joinDelimiter() {
+        return this.joinDelimiter;
+    }
+
+    @Override
+    public DocumentSplitter defaultSubSplitter() {
+        return null;
```

Comment:
I do think you might want to specify a range. For example, only split on H1 and H2 and then only paragraphs. I might want this because I don't think the other headers, new lines, sentences, or words bring semantic meaning and I would prefer not to split them. So the split would stop at paragraphs - either you can fit one more in the context or not but don't split finer.
Does that make sense?

Reply:
I think my request would also touch this part of the code
https://github.com/langchain4j/langchain4j/blob/e9a8c25df171f8fd0a16f799baa6d9439a4c3d3b/langchain4j/src/main/java/dev/langchain4j/data/document/splitter/HierarchicalDocumentSplitter.java#L147

Reply:
@thesteve0 having it configurable seems good, if it does not bring a lot of complexity